src/testing/slim_check/testable.lean

```diff
@@ -0,0 +1,490 @@
+io.run_rand (testable.run_suite p cfg)
+
+namespace tactic
+
+open tactic expr
+
+/-!
+## Decorations
+
+Instances of `testable` use `named_binder` as a decoration on
+propositions in order to access the name of bound variables, as in
+`named_binder (some "x") (forall x, x < y)`.  This helps the
+`testable` instances create useful error messages where variables
+are matched with values that falsify a given proposition.
+
+The following functions help support the gadget so that the user does
+not have to put them in themselves.
+-/
+
+/-- `add_existential_decorations p` adds `a `named_binder` annotation at the
+root of `p` if `p` is an existential quantification -/
+meta def add_existential_decorations : expr → expr
+| e@`(@Exists %%α %%(lam n bi d b)) :=
+  let n := to_string n in
+  const ``named_binder [] (const ``option.some [level.zero] (`(string) : expr) (`(n) : expr)) e
+| e := e
+
+/-- Traverse the syntax of a proposition to find universal quantifiers
+and existential quantifiers and add `named_binder` annotations next to
+them. -/
+meta def add_decorations : expr → expr | e :=
+e.replace $ λ e _,
+  match e with
+  | (pi n bi d b) :=
+    let n := to_string n in
+    some $ const ``named_binder [] (const ``option.some [level.zero] (`(string) : expr) (reflect n : expr)) (pi n bi (add_existential_decorations d) (add_decorations b))
+  | e := none
+  end
+
+/-- `decorations_of p` is used as a hint to `mk_decorations` to specify
+that the goal should be satisfied with a proposition equivalent to `p`
+with added annotations. -/
+@[reducible, nolint unused_arguments]
+def decorations_of (p : Prop) := Prop
+
+/-- In a goal of the shape `⊢ tactic.decorations_of p`, `mk_decoration` examines
+the syntax of `p` and add `named_binder` around universal quantifications and
+existential quantifications to improve error messages.
+
+This tool can be used in the declaration of a function as follows:
+
+```lean
+def foo (p : Prop) (p' : tactic.decorations_of p . mk_decorations) [testable p'] : ...
+```
+
+`p` is the parameter given by the user, `p'` is an equivalent proposition where
+the quantifiers are annotated with `named_binder`.
+-/
+meta def mk_decorations : tactic unit := do
+`(tactic.decorations_of %%p) ← target,
+exact $ add_decorations p
+
+end tactic
+
+/-- Run a test suite for `p` and return true or false: should we believe that `p` holds? -/
+def testable.check (p : Prop) (cfg : slim_check_cfg := {}) (p' : tactic.decorations_of p . tactic.mk_decorations) [testable p'] : io bool := do
+x ← io.run_rand (testable.run_suite p' cfg),
+match x with
+ | (success _) := io.put_str_ln ("Success") >> return tt
+ | (gave_up n) := io.put_str_ln ("Gave up " ++ repr n ++ " times") >> return ff
+ | (failure _ xs) := do
+   io.put_str_ln "\n===================",
+   io.put_str_ln "Found problems!",
+   io.put_str_ln "",
+   list.mmap' io.put_str_ln xs,
+   io.put_str_ln "-------------------",
+   return ff
+end
+
```

Comment:
I'm surprised to see this end without a `slim_check` tactic! It feels kind of like it's missing a punchline. It should be easy to port the io version to a tactic `slim_check` that tests the goal (possibly reverting hypotheses?) and a command `#slim_check`, right?

Reply:
Indeed and it is in the bigger version of `slim_check`. I tried to break off a chewable piece for a review. I think it might be good to look at the bells and whistles of the `slim_check` tactic separately from the sampling infrastructure.

Reply:
Okay, sure.